internal/controller/utils.go

```diff
@@ -522,3 +556,124 @@ func healthCheck(sts *appsv1.StatefulSet, lg klog.Logger) (*clientv3.MemberListR
 
 	return memberlistResp, healthInfos, nil
 }
+
+func createCMCertificateConfig(ec *ecv1alpha1.EtcdCluster) *certInterface.Config {
+	cmConfig := ec.Spec.TLS.ProviderCfg.CertManagerCfg
+	duration, err := time.ParseDuration(cmConfig.ValidityDuration)
+	if err != nil {
+		log.Printf("Failed to parse ValidityDuration: %s", err)
+	}
+
+	var getAltNames certInterface.AltNames
+	if cmConfig.AltNames.DNSNames != nil {
+		getAltNames = certInterface.AltNames{
+			DNSNames: cmConfig.AltNames.DNSNames,
+			IPs:      make([]net.IP, len(cmConfig.AltNames.DNSNames)),
+		}
+	} else {
+		defaultDNSNames := []string{fmt.Sprintf("%s.svc.cluster.local", cmConfig.CommonName)}
+		getAltNames = certInterface.AltNames{
+			DNSNames: defaultDNSNames,
+		}
+	}
+
+	config := &certInterface.Config{
+		CommonName:       cmConfig.CommonName,
+		Organization:     cmConfig.Organization,
+		ValidityDuration: duration,
+		AltNames:         getAltNames,
+		ExtraConfig: map[string]any{
+			"issuerName": cmConfig.IssuerName,
+			"issuerKind": cmConfig.IssuerKind,
+		},
+	}
+	return config
+}
+
+func createAutoCertificateConfig(ec *ecv1alpha1.EtcdCluster) *certInterface.Config {
+	// TODO
+	config := &certInterface.Config{}
+	return config
+}
+
+func createCertificate(ec *ecv1alpha1.EtcdCluster, ctx context.Context, c client.Client, certName string) error {
+	cert, certErr := certificate.NewProvider(certificate.ProviderType(ec.Spec.TLS.Provider), c)
+	if certErr != nil {
+		// TODO: instead of error, set default autoConfig
+		return certErr
+	}
+	_, getCertError := cert.GetCertificateConfig(ctx, certName, ec.Namespace)
+	if getCertError != nil {
+		if k8serrors.IsNotFound(getCertError) {
+			log.Printf("Creating certificate: %s for etcd-operator: %s\n", certName, ec.Name)
+			switch {
+			case ec.Spec.TLS.ProviderCfg.AutoCfg != nil:
+				cmConfig := createAutoCertificateConfig(ec)
+				createCertErr := cert.EnsureCertificateSecret(ctx, certName, ec.Namespace, cmConfig)
+				if createCertErr != nil {
+					log.Printf("Error creating certificate: %s", createCertErr)
+				}
+				return nil
+			case ec.Spec.TLS.ProviderCfg.CertManagerCfg != nil:
+				cmConfig := createCMCertificateConfig(ec)
+				createCertErr := cert.EnsureCertificateSecret(ctx, certName, ec.Namespace, cmConfig)
+				if createCertErr != nil {
+					log.Printf("Error creating certificate: %s", createCertErr)
+				}
+				return nil
+			default:
+				// TODO: Use AuthProvider, since both AutoCfg and CertManagerCfg is not present
+				log.Printf("Error creating certificate, valid certificate provider not defined.")
+				return nil
+			}
+		} else {
+			return fmt.Errorf("%s:Error getting certificate", getCertError)
+		}
+	}
+
+	return nil
+}
+
+func createClientCertificate(ec *ecv1alpha1.EtcdCluster, ctx context.Context, c client.Client) error {
+	certName := fmt.Sprintf("%s-%s-tls", ec.Name, "client")
```

Comment:
create a function to get the secret for client

Reply:
Added a `getClientCertName` function to generate the Client Certificate name and since we are keeping the certificate and the secret name same, we can reuse it to get the secret name.